sdks/java/io/xml/src/test/java/org/apache/beam/sdk/io/xml/XmlSourceTest.java

```diff
@@ -873,6 +881,46 @@ public void testSplitAtFractionExhaustiveSingleByte() throws Exception {
     assertSplitAtFractionExhaustive(source, options);
   }
 
+  @Test
+  public void testNoBufferOverflowThrown() throws IOException {
+    // The magicNumber was found imperatively and will be different for different xml content.
+    // Test with the current setup causes BufferOverflow in
+    // XMLReader#getFirstOccurenceOfRecordElement method,
+    // if the specific corner case is not handled
+    final int magicNumber = 183;
+    StringBuilder sb = new StringBuilder();
```

Comment:
I think your input here does not conform to the format required by the Xml source that is defined here: https://github.com/apache/beam/blob/master/sdks/java/io/xml/src/main/java/org/apache/beam/sdk/io/xml/XmlIO.java#L68

Specifically it has to be of the following format.
   * <root>
   * <record> ... </record>
   * <record> ... </record>
   * <record> ... </record>
   * ...
   * <record> ... </record>
   * </root>
But you have additional element: <trainTags><trainTag></trainTag> ... <trainTag></trainTag></trainTags>

Were you able to reproduce the issue when the input conforms to the required format ?

Reply:
This is an artificial example, but the main condition is that you have inner `<recordBlahBlah>` tags inside the `record`
The real case looks like
```
<root>
 <record> 
  <recordSomething>
  </recordSomething>
 </record>
 <record> 
  <recordSomething>
  </recordSomething>
 </record>
...
 <record> 
  <recordSomething>
  </recordSomething>
 </record>
</root>
```
The behaviour seems to be environment-dependent, so I'm not sure if even my example will be reproduced in another environment (it depends also on how many bytes we are reading  from the channel)

Reply:
Here is a piece of XML we deal with in our service:
from the root `ONIXMessage` we read records `Product`
```
<?xml version='1.0' encoding='ISO-8859-1'?>
<ONIXMessage xmlns="http://ns.editeur.org/onix/3.0/reference" release="3.0">
  <Header>
    <Sender>
      <SenderIdentifier>
        <SenderIDType></SenderIDType>
        <IDValue></IDValue>
      </SenderIdentifier>
      <SenderName></SenderName>
      <EmailAddress></EmailAddress>
    </Sender>
    <Addressee>
      <AddresseeName></AddresseeName>
    </Addressee>
    <MessageNumber></MessageNumber>
    <SentDateTime></SentDateTime>
  </Header>
  <Product>
    <RecordReference></RecordReference>
    <NotificationType></NotificationType>
    <RecordSourceType></RecordSourceType>
    <ProductIdentifier>
      <ProductIDType></ProductIDType>
      <IDValue></IDValue>
    </ProductIdentifier>
    <ProductIdentifier>
      <ProductIDType></ProductIDType>
      <IDValue></IDValue>
    </ProductIdentifier>
```
as you can see there are inner tags like `ProductIdentifier` and `ProductIDType`.
But we also have structore like
```
<root>
<head>
</head>
<record>
</record>
...
</root>
```
and _in most cases_, it's not a problem
We run it in Google Dataflow, and when it fails with BufferOverflow restarting the job helps - so the issue can't be easily reproduced.

Reply:
I've updated the test - now it uses input like
```
<trains>
    <train>
        <trainTags>
            <trainTag>0</trainTag>
            <trainTag>1</trainTag>
            <trainTag>2</trainTag>
            ...
 ```
 So the format issue is not the case.
 
To reproduce the error I had to use TestPipeline - so the input was split into bundles
> INFO: Splitting filepattern /var/folders/j5/2qx0r7453tvd56zpjbstv6fw4m_zm3/T/junit946409719265813413/trainXMLWithTags into bundles of size 126 took 1 ms and produced 1 files and 20 bundles